```{r}
# Necessary libraries
library(Seurat)
library(cowplot)
library(ggplot2)
library(patchwork)
library(dplyr)
library(RColorBrewer)
library(httpgd)
library(stringr)
library(stringi)
library(limma)
library(readxl)
library(ggfortify)
library(reshape2)
library(factoextra)
library(vegan)
# Prepare for Cell Cycle Regression
s_genes <- cc.genes$s.genes
g2m_genes <- cc.genes$g2m.genes
```

```{r}
# Determine Drive location
if (Sys.info()["sysname"] == "Windows") {
google_drive <- "G:"
} else if (Sys.info()["sysname"] == "Darwin") {
google_drive <-
"/Users/robinyan/Library/CloudStorage/GoogleDrive-robinyzx@umich.edu"
}

# Data Import
current <- getwd()

# Load PASE data
setwd(file.path(google_drive,
"My Drive/Projects/Yi's scRNA Data/Data/Processed"))
load("allsamples_PCA_cluster_dims20_res0.5.Rd")
pase <- allsamples_PCA_cluster_dims20_res0.5

# Load Human Gastrula
setwd(file.path(google_drive,
"My Drive/Projects/Yi's scRNA Data/External Papers/Human gastrula"))
gastrula_mat <- readRDS("raw_matrix.rds")
gastrula_anno <- readRDS("annot_umap.rds")

# Load Human E9E11
setwd(file.path(google_drive,
"My Drive/Projects/Yi's scRNA Data/External Papers/Human E9E11"))
e9e11 <- readRDS("embryo_integrated_allembryos_filtered.Rdata")

# Load Preimplantation data from Petropoulos
setwd(file.path(google_drive,
"My Drive/Projects/Yi's scRNA Data/External Papers/Petropoulos"))
load("Preimplatation_UMAP_OrigID.Rd")
preimp <- Preimplatation_PCA_cluster_dims20_res0.2

# Load Dish, 0h and 12h PASE
masterdir <- file.path(google_drive,
"My Drive/Projects/Yi's scRNA Data/Data")

# Dish H9
load(file.path(masterdir, "New_2022_04", "DishH9.Rd"))
h9_dish <- scdata_PCA_cluster_dims20_res0.5

# 0h
load(file.path(masterdir, "New_2022_04", "0h_H9.Rd"))
pase_0h <- scdata_PCA_cluster_dims20_res0.5

# 12h
load(file.path(masterdir, "New_2022_04", "uPASE_12h.Rd"))
pase_12h <- scdata_PCA_cluster_dims20_res0.5

setwd(current)
```

```{r}
# Pre-processing - Human E9E11
# Update Seurat object
e9e11 <- UpdateSeuratObject(object = e9e11)

# Save current clustering idents
e9e11$ident <- e9e11@active.ident

# Convert embryo info into embryonic day info
e9e11_day <- e9e11$orig.ident
e9e11_day <- gsub("embryo3", "Day 11", e9e11_day, fixed = TRUE)
e9e11_day <- gsub("embryo4", "Day 11", e9e11_day, fixed = TRUE)
e9e11_day <- gsub("embryo9", "Day 11", e9e11_day, fixed = TRUE)
e9e11_day <- gsub("embryo10", "Day 11", e9e11_day, fixed = TRUE)
e9e11_day <- gsub("embryo13", "Day 11", e9e11_day, fixed = TRUE)
e9e11_day <- gsub("embryo15", "Day 11", e9e11_day, fixed = TRUE)
e9e11_day <- gsub("embryo2", "Day 9", e9e11_day, fixed = TRUE)
e9e11_day <- gsub("embryo7", "Day 9", e9e11_day, fixed = TRUE)
e9e11_day <- gsub("embryo12", "Day 9", e9e11_day, fixed = TRUE)
e9e11_day <- gsub("embryo14", "Day 9", e9e11_day, fixed = TRUE)
e9e11_day <- gsub("embryo1", "Day 9", e9e11_day, fixed = TRUE)
e9e11$day <- e9e11_day

# Cell Cycle Score
e9e11 <- CellCycleScoring(e9e11, s.features = s_genes,
g2m.features = g2m_genes, set.ident = TRUE)
e9e11[["CC.Difference"]] <- e9e11$S.Score - e9e11$G2M.Score

# Subset E9E11 into E9 and E11
e9 <- subset(x = e9e11, subset = day == "Day 9")
e11 <- subset(x = e9e11, subset = day == "Day 11")

# Save cluster ident for use after integration
e9$e9_ident <- e9$ident
e11$e11_ident <- e11$ident
```

```{r}
# Pre-processing - PASE
# Save PASE ident for use after integration
pase$pase_ident <- pase@active.ident
```

```{r}
# Pre-processing - Human Gastrula
# Normalize gastrula data to 10X compatible format
tx_size <- read.table("Homo_sapiens.GRCh38.102.gtf.exonsSizePerGene.txt",
header = T)
tx_size_uniq <- tx_size[!duplicated(tx_size$gene_name), ]
raw <- as.data.frame(t(gastrula_mat))
raw[, "gene_name"] <- rownames(raw)
raw_matrix_txlen <- merge(tx_size_uniq, raw, by = "gene_name")
raw_matrix_txnorm <- as.matrix(raw_matrix_txlen[, -1:-3])
class(raw_matrix_txnorm) <- "numeric"
raw_matrix_txnorm <- raw_matrix_txnorm * 1000 / raw_matrix_txlen[, 3]
rownames(raw_matrix_txnorm) <- as.character(raw_matrix_txlen[, 1])

# Standard Seurat workflow
gastrula <- CreateSeuratObject(counts = raw_matrix_txnorm)
gastrula <- NormalizeData(object = gastrula)
gastrula <- FindVariableFeatures(object = gastrula)
# Add metadata on cluster identity
gastrula$gastrula_ident <- gastrula_anno$sub_cluster

# Cell Cycle Score
gastrula <- CellCycleScoring(gastrula, s.features = s_genes,
g2m.features = g2m_genes, set.ident = TRUE)
gastrula[["CC.Difference"]] <- gastrula$S.Score - gastrula$G2M.Score

# Select relevant clusters to create gastrula_selected
Idents(object = gastrula) <- "gastrula_ident"
gastrula_selected <-
subset(gastrula, idents = c("Primitive Streak",
"Emergent Mesoderm", "Epiblast", "Nascent Mesoderm",
"Non-Neural Ectoderm", "PGC"))

# Remove blood clusters to create gastrula_no_blood
gastrula_no_blood <-
subset(gastrula, idents = c("Advanced Mesoderm",
"Axial Mesoderm", "DE(NP)", "DE(P)", "Emergent Mesoderm",
"Epiblast", "Hypoblast", "Nascent Mesoderm", "Non-Neural Ectoderm",
"PGC", "Primitive Streak", "YS Endoderm", "YS Mesoderm"))
```

```{r}
# Pre-processing - Preimp
# Rename NA as Morula
cluster_id <- c("Trophectoderm", "Morula", "Preimplantation Epiblast",
"Primitive Endoderm")
names(cluster_id) <- levels(preimp)
preimp <- RenameIdents(preimp, cluster_id)

# Save PASE ident for use after integration
preimp$preimp_ident <- preimp@active.ident
```







```{r}
# Downsample PASE to 100 cells each cluster (800 in total)
Idents(pase) <- pase$pase_ident
pase_downsampled <- subset(x = pase, downsample = 100)
```

```{r}
# Downsample Preimp to 100 cells each cluster (400 in total)
Idents(preimp) <- preimp$preimp_ident
preimp_downsampled <- subset(x = preimp, downsample = 100)
```

```{r}
# Downsample E9, E11 to 100 cells each cluster (400 in total each)
Idents(e9) <- e9$e9_ident
Idents(e11) <- e11$e11_ident
e9_downsampled <- subset(x = e9, downsample = 100)
e11_downsampled <- subset(x = e11, downsample = 100)
```

```{r}
# Downsample Dish H9, 0h PASE and 12h PASE
h9_dish_downsampled <- subset(x = h9_dish, downsample = 100)
pase_0h_downsampled <- subset(x = pase_0h, downsample = 100)
pase_12h_downsampled <- subset(x = pase_12h, downsample = 100)
```










```{r}
# Integration - downsampled PASE and gastrula all
# Combine two Seurat objects into list
combined <- list("pase" = pase_downsampled, "gastrula" = gastrula)
# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = combined)
anchors <- FindIntegrationAnchors(object.list = combined,
anchor.features = features)
integrated <- IntegrateData(anchorset = anchors)
# Integrated Analysis
DefaultAssay(integrated) <- "integrated"
integrated <- ScaleData(integrated,
vars.to.regress = "CC.Difference", features = rownames(integrated))
integrated <- RunPCA(integrated, npcs = 20, verbose = FALSE)
integrated <- RunUMAP(integrated, reduction = "pca", dims = 1:20,
return.model = TRUE)
```

```{r}
# Export
pdf(file = "pase_gastrula_all_paseID.pdf", width = 8, height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "pase_ident")
g <- g & theme(legend.text = element_blank())
g
dev.off()
pdf(file = "pase_gastrula_all_paseID_with_names.pdf", width = 8, height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "pase_ident")
g
dev.off()
pdf(file = "pase_gastrula_all_gastrulaID.pdf", width = 8, height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "gastrula_ident")
g <- g & theme(legend.text = element_blank())
g
dev.off()
pdf(file = "pase_gastrula_all_gastrulaID_with_names.pdf", width = 8, height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "gastrula_ident")
g
dev.off()
```














```{r}
# Integration - downsampled PASE and gastrula selected
# Combine two Seurat objects into list
combined <- list("pase" = pase_downsampled, "gastrula" = gastrula_selected)
# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = combined)
anchors <- FindIntegrationAnchors(object.list = combined,
anchor.features = features)
integrated <- IntegrateData(anchorset = anchors)
# Integrated Analysis
DefaultAssay(integrated) <- "integrated"
integrated <- ScaleData(integrated,
vars.to.regress = "CC.Difference", features = rownames(integrated))
integrated <- RunPCA(integrated, npcs = 20, verbose = FALSE)
integrated <- RunUMAP(integrated, reduction = "pca", dims = 1:20,
return.model = TRUE)
```

```{r}
# Export
pdf(file = "pase_gastrula_selected_paseID.pdf", width = 8, height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "pase_ident")
g <- g & theme(legend.text = element_blank())
g
dev.off()
pdf(file = "pase_gastrula_selected_paseID_with_names.pdf", width = 8,
height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "pase_ident")
g
dev.off()
pdf(file = "pase_gastrula_selected_gastrulaID.pdf", width = 8, height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "gastrula_ident")
g <- g & theme(legend.text = element_blank())
g
dev.off()
pdf(file = "pase_gastrula_selected_gastrulaID_with_names.pdf", width = 8,
height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "gastrula_ident")
g
dev.off()
```
















```{r}
# Integration - downsampled PASE, gastrula no blood, downsampled e9&e11
# Combine two Seurat objects into list
combined <- list("pase" = pase_downsampled, "gastrula" = gastrula,
"e9" = e9_downsampled, "e11" = e11_downsampled)
# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = combined)
anchors <- FindIntegrationAnchors(object.list = combined,
anchor.features = features)
integrated <- IntegrateData(anchorset = anchors)
# Integrated Analysis
DefaultAssay(integrated) <- "integrated"
integrated <- ScaleData(integrated,
vars.to.regress = "CC.Difference", features = rownames(integrated))
integrated <- RunPCA(integrated, npcs = 20, verbose = FALSE)
integrated <- RunUMAP(integrated, reduction = "pca", dims = 1:20,
return.model = TRUE)
```

```{r}
# Visualization
DimPlot(integrated, reduction = "umap", group.by = "pase_ident")
DimPlot(integrated, reduction = "umap", group.by = "gastrula_ident")
DimPlot(integrated, reduction = "umap", group.by = "e9_ident")
DimPlot(integrated, reduction = "umap", group.by = "e11_ident")
```

```{r}
# Export
pdf(file = "pase_gastrula_all_e9e11_paseID.pdf", width = 8,
height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "pase_ident")
g <- g & theme(legend.text = element_blank())
g
dev.off()
pdf(file = "pase_gastrula_all_e9e11_gastrulaID.pdf", width = 8,
height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "gastrula_ident")
g <- g & theme(legend.text = element_blank())
g
dev.off()
pdf(file = "pase_gastrula_all_e9e11_e9ID.pdf", width = 8,
height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "e9_ident")
g <- g & theme(legend.text = element_blank())
g
dev.off()
pdf(file = "pase_gastrula_all_e9e11_e11ID.pdf", width = 8,
height = 8)
g <- DimPlot(integrated, reduction = "umap", group.by = "e11_ident")
g <- g & theme(legend.text = element_blank())
g
dev.off()
```



































```{r}
# Take out epiblast from downsampled data
gastrula_epi <- subset(x = gastrula, idents = c("Epiblast"))
gastrula_epi$source <- "Gastrula" # add label for later reference

pase_epi <- subset(x = pase_downsampled, subset = pase_ident == "ESC")
pase_epi$source <- "PASE"

h9_dish_downsampled$source <- "Dish H9"
pase_0h_downsampled$source <- "PASE 0h"
pase_12h_downsampled$source <- "PASE 12h"

# Follwing contains more than one cluster
e9e11_epi <- subset(x = e9e11, subset = ident == "Epiblast") #E9E11 Epi is small
e9e11_epi$e9e11_ident <- "Epiblast"

preimp_epi <- subset(x = preimp_downsampled,
idents = c("Preimplantation Epiblast", "Morula"))
```







```{r}
# Epi gene list
epi_gene_list <- as.list(read_excel("epi_gene_list.xlsx"))
epi_gene_list <- epi_gene_list[[1]]
```

```{r}
# Integration
combined <- list("pase" = pase_epi, "gastrula" = gastrula_epi,
"e9e11" = e9e11_epi, "preimp" = preimp_epi, "dish_h9" = h9_dish_downsampled,
"pase_0h" = pase_0h_downsampled, "pase_12h" = pase_12h_downsampled)
# Use epi gene list as highly variable genes
present_list <- epi_gene_list
# Only genes present in all samples
present_list <- intersect(present_list, rownames(pase_epi))
present_list <- intersect(present_list, rownames(gastrula_epi))
present_list <- intersect(present_list, rownames(e9e11_epi))
present_list <- intersect(present_list, rownames(preimp_epi))
present_list <- intersect(present_list, rownames(h9_dish_downsampled))
present_list <- intersect(present_list, rownames(pase_0h_downsampled))
present_list <- intersect(present_list, rownames(pase_12h_downsampled))
features <- present_list

anchors <- FindIntegrationAnchors(object.list = combined,
anchor.features = features)
integrated <- IntegrateData(anchorset = anchors)
```

```{r}
# Direct PCA within Seurat
# Combine representative metadata
integrated$cell_source <- coalesce(integrated$source, integrated$day,
integrated$preimp_ident)

# Extract raw count data
merged_counts <- integrated@assays$RNA@counts
merged_counts_epi_gene <- merged_counts[features, ]

# Extract normalized data
merged_data <- integrated@assays$RNA@data
merged_data_epi_gene <- merged_data[features, ]

# Seurat workflow without normalization
merged <- CreateSeuratObject(counts = merged_counts_epi_gene)
merged@assays$RNA@data <- merged_data_epi_gene # use previously normalized
merged <- FindVariableFeatures(object = merged)
merged <- ScaleData(object = merged)
merged <- RunPCA(object = merged)
merged$cell_source <- integrated$cell_source # pass on meta data

# Visualization
DimPlot(merged, reduction = "pca", group.by = "cell_source", dims = c(1, 3))

# Calculate variance explained in PCA
mat <- Seurat::GetAssayData(merged, assay = "RNA", slot = "scale.data")
pca <- merged[["pca"]]
# Get the total variance:
total_variance <- sum(matrixStats::rowVars(mat))
pca_variance <- (pca@stdev)^2  ## EigenValues
var_explained <- pca_variance / total_variance
var_explained
```

```{r}
# Export
pdf(file = "all_epi_genes_seurat_pca_dim13.pdf", width = 8,
height = 8)
g <-
DimPlot(merged, reduction = "pca", group.by = "cell_source", dims = c(1, 3))
g <- g & theme(legend.text = element_blank())
g
dev.off()
```

```{r}
# install.packages('plotly')

# Load plot_ly
# library(plotly)

# Construct a dataframe using data from your pre-clustered Seurat v3.1.1 object
# Here 'seurat_clusters' is list of numeric cluster identities, you can find it here: yourseuratobject[["seurat_cluster"]], 
# or yourseuratobject$seurat_clusters, where 'yourseuratobject' is a Seurat object created with Seurat v3.1.1 (works for v3.0.0 as well)
yourseuratobject <- merged

# Re-run UMAPs that you have accurate calculations for all UMAP(s)
# yourseuratobject <- RunUMAP(yourseuratobject,
#                             dims = 1:10,
#                             n.components = 3L)

# This is a manual method of extracting embeddings and is not needed
# as pointed out by user @sdinardo on 01142022 thank you! 
# Extract UMAP information from Seurat Object
# UMAP_1 <- yourseuratobject[["umap"]]@cell.embeddings[,1]
# UMAP_2 <- yourseuratobject[["umap"]]@cell.embeddings[,2]
# UMAP_3 <- yourseuratobject[["umap"]]@cell.embeddings[,3]

# Visualize what headings are called so that you can extract them to form a dataframe
Embeddings(object = yourseuratobject, reduction = "pca")

# Prepare a dataframe for cell plotting
plot.data <- FetchData(object = yourseuratobject, vars = c("PC_1", "PC_2", "PC_3", "cell_source"))

# Make a column of row name identities (these will be your cell/barcode names)
plot.data$label <- paste(rownames(plot.data))

# Plot your data, in this example my Seurat object had 21 clusters (0-20)
fig <- plot_ly(data = plot.data, 
        x = ~PC_1, y = ~PC_2, z = ~PC_3, 
        color = ~cell_source, 
        colors = c("#F68282",
                   "#cd9600",
                   "#7cae00",
                   "#0cb702",
                   "#00c19a",
                   "#00b8e7",
                   "#00a9ff",
                   "#8494ff",
                   "#ff68a1"),
        type = "scatter3d", 
        mode = "markers", 
        marker = list(size = 3, width=2), # controls size of points
        text=~label, #This is that extra column we made earlier for which we will use for cell ID
        hoverinfo="text") #When you visualize your plotly object, hovering your mouse pointer over a point shows cell names


# Updates stemming from Issue #9 Having a fixed scale on axes while selecting particular clusters
# @rtoddler thanks for the suggestions!
# Before you plot, set the ranges of the axis you desire. This set axis range will be 
# present across all clusters, and plotly will not adjust for axis length anymore
# this axis length will persist even when selecting some clusters

# xaxis
axx <- list(
  nticks = 4,
  range = c(-15,25) #select range of xaxis
)

# yaxis
axy <- list(
  nticks = 4,
  range = c(-20,20) #select range of yaxis
)

#zaxis
axz <- list(
  nticks = 4,
  range = c(-15,15) #select range of zaxis
)

fig <- fig %>% layout(scene = list(xaxis=axx,yaxis=axy,zaxis=axz))
fig_cube <- fig %>% layout(scene = list(xaxis=axx,yaxis=axy,zaxis=axz, aspectmode='cube')) # To maintain cubic aspect
fig
fig_cube
```

```{r}
# Say you wanto make a gene-expression 3D plot, where you can plot gene expression against a color scale
# Here using the same seurat object as above, we extract gene expression information for beta-actin 'ACTB'
# Here we concentrate on SCT normalized data, or log normalized RNA NOT raw counts.
# In addition if you want, you may look at normalised-RNA, SCT or integrated slots, to look at gene expression
# Setting your DefaultAssay() will inform R which assay to pick up expression data from.
DefaultAssay(object = yourseuratobject)
DefaultAssay(object = yourseuratobject) <- "RNA"
DefaultAssay(object = yourseuratobject) <- "integrated"
DefaultAssay(object = yourseuratobject) <- "SCT"

# create a dataframe
plot.data <- FetchData(object = yourseuratobject, vars = c("UMAP_1", "UMAP_2", "UMAP_3", "ACTB"), slot = 'data')

# Say you want change the scale, so that every cell having an expression >1 will be one color
# Basically, you are re-adjusting the scale here, so that any cell having a certain expression will light up on your 3D plot

# First make another column in your dataframe, where all values above 1 are re-assigned a value of 1
# This information is stored in the 'changed' column of your dataframe
plot.data$changed <- ifelse(test = plot.data$ACTB <1, yes = plot.data$ACTB, no = 1)

# Add the label column, so that now the column has 'cellname-its expression value'
plot.data$label <- paste(rownames(plot.data)," - ", plot.data$ACTB, sep="")

# Plot your data, in this example my Seurat object had 21 clusters (0-20), and cells express a gene called ACTB
plot_ly(data = plot.data, 
        x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3, 
        color = ~changed, # you can just run this against the column for the gene as well using ~ACTB, the algorith will automatically scale in that case based on maximal and minimal values
        opacity = .5,
        colors = c('darkgreen', 'red'), 
        type = "scatter3d", 
        mode = "markers",
        marker = list(size = 5, width=2), 
        text=~label,
        hoverinfo="text"
)

# On running this code the HTML output should appear in RStudio. You can save the output as a
# HTML file. Once you have saved, just open the HTML file in any web browser (double click on the html- file
# and if asked select to open with any web browser like google chrome/safari/mozilla/explorer etc).
# It should be have all of the integrated features you saw in the RStudio output file.

########## #
########## #

# Alternative method as designed by @vertesy (Thanks for the suggestions!)
# create a dataframe
goi <- "TOP2A"
plotting.data <- FetchData(object = yourseuratobject, vars = c("UMAP_1", "UMAP_2", "UMAP_3", "Expression"=goi), slot = 'data')

# Say you want change the scale, so that every cell having an expression >1 will be one color
# Basically, you are re-adjusting the scale here, so that any cell having a certain expression will light up on your 3D plot

# First make another column in your dataframe, where all values above 1 are re-assigned a value of 1
# This information is stored in the 'Expression' column of your dataframe
# Cutoff <- 2
Cutoff <- quantile(plotting.data[,goi], probs = .95)
plotting.data$"ExprCutoff" <- ifelse(test = plotting.data[,goi] <Cutoff, yes = plotting.data[,goi], no = Cutoff)

# Add the label column, so that now the column has 'cellname-its expression value'
plotting.data$label <- paste(rownames(plotting.data)," - ", plotting.data[,goi], sep="")

# Plot your data, in this example my Seurat object had 21 clusters (0-20), and cells express a gene called ACTB
plot_ly(data = plotting.data,
        # name = goi,
        x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3, 
        color = ~ExprCutoff, # you can just run this against the column for the gene as well using ~ACTB, the algorith will automatically scale in that case based on maximal and minimal values
        opacity = .5,
        colors = c('darkgrey', 'red'), 
        type = "scatter3d", 
        mode = "markers",
        marker = list(size = 1), 
        text=~label,
        hoverinfo="text"
) %>%layout(title=goi)
```



















```{r}
# Heatmap
# E9E11 rebuild
day <- e9e11_epi$day
e9e11_epi <- CreateSeuratObject(counts = e9e11_epi@assays$RNA@counts)
e9e11_epi$day <- day

# Integration
combined <- list("pase" = pase_epi, "gastrula" = gastrula_epi,
"e9e11" = e9e11_epi, "preimp" = preimp_epi, "dish_h9" = h9_dish_downsampled,
"pase_0h" = pase_0h_downsampled, "pase_12h" = pase_12h_downsampled)

# Separate normalization to CPM natural log1p
combined <- lapply(X = combined, FUN = function(x) {
    x <- NormalizeData(x, normalization.method = "LogNormalize",
    scale.factor = 1e6)
})

anchors <- FindIntegrationAnchors(object.list = combined)
integrated <- IntegrateData(anchorset = anchors)

# Combine representative metadata
integrated$cell_source <- coalesce(integrated$source, integrated$day,
integrated$preimp_ident)

# Gene list
# heatmap_genes <- c("NANOG", "SOX2", "POU5F1", "KLF2", "KLF4", "TFCP2L1",
# "DPPA3", "ESRRB", "TBX3", "OTX2", "POU3F1", "SOX4", "DNMT3B", "DNMT3A",
# "DNMT3L", "FGF4", "FGF5", "ZIC2", "LIN28B", "ESRP1", "ETV1", "ETV4", "GRHL2",
# "CLDN6", "CLDN7", "TBXT", "TFAP2A", "ACTC1", "CER1", "EOMES", "FOXA2", "CTNNB1",
# "NODAL", "BMP2", "BMP4")

heatmap_genes <- rev(c("POU5F1", "SOX2", "NANOG", "KLF2", "KLF4", "KLF17",
"TFCP2L1", "DPPA2", "DPPA3", "DPPA5", "TFAP2C", "SOX15", "CD53", "ESRRB",
"TBX3",
"DNMT3A", "DNMT3B", "DNMT3L", "OTX2", "POU3F1", "SOX1", "SOX3", "SOX4",
"FGF2",
"FGF4", "FGF5", "ZIC2", "LIN28B", "ESRP1", "ETV1", "ETV4", "GRHL2", "CLDN6",
"CLDN7", "ZNF676", "ZNF560", "ZNF729", "ZNF728", "TFAP2A", "TBXT",
"ACTC1", "CER1", "EOMES", "FOXA2", "CTNNB1", "NODAL", "BMP2", "BMP4",
"GATA6", "ID2"))

# Average expression in each cluster (exponentiate "data")
heatmap_data <- AverageExpression(integrated, assays = "RNA",
group.by = "cell_source", slot = "data", features = heatmap_genes)

# Convert to 2-based log1p
heatmap_data <- log2(heatmap_data$RNA + 1)

# Rearrange columns
col_order <- c("Morula", "Preimplantation Epiblast", "Day 9",
"Day 11", "Dish H9", "PASE 0h", "PASE 12h", "PASE", "Gastrula")
heatmap_data <- heatmap_data[, col_order]


# Visualization

# Heatmap Visualization
melted <- melt(heatmap_data)
heatmap <- ggplot(data = melted, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile() + scale_fill_distiller(palette = "YlOrBr") +
  coord_flip() + theme(axis.title.x = element_blank(),
  axis.title.y = element_blank())
heatmap


# Plot modification
# Modify elements
heatmap <- heatmap + theme(
axis.ticks = element_line(size = 0.4),
axis.line = element_line(size = 0.4),
axis.text = element_text(size = 5, face = NULL),
axis.text.x = element_text(angle = 90),
legend.text = element_text(size = 5, face = NULL),
legend.title = element_blank(),
legend.key.width = unit(0.1, "in"))

pdf(file = "heatmap.pdf", width = 3, height = 4)
heatmap
dev.off()
```





















































```{r}
# Integration - downsampled PASE and monkey
# Combine two Seurat objects into list
combined <- list("pase" = pase_downsampled, "monkey" = monkey)
# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = combined)
anchors <- FindIntegrationAnchors(object.list = combined,
anchor.features = features)
integrated <- IntegrateData(anchorset = anchors)
# Integrated Analysis
DefaultAssay(integrated) <- "integrated"
integrated <- ScaleData(integrated,
vars.to.regress = "CC.Difference", features = rownames(integrated))
integrated <- RunPCA(integrated, npcs = 20, verbose = FALSE)
integrated <- RunUMAP(integrated, reduction = "pca", dims = 1:20,
return.model = TRUE)
```

```{r}
# Export
svg(file = "pase_monkey_paseID.svg", width = 8, height = 8)
DimPlot(integrated, reduction = "umap", group.by = "pase_ident")
dev.off()
svg(file = "pase_monkey_selected_gastrulaID.svg", width = 8, height = 8)
DimPlot(integrated, reduction = "umap", group.by = "monkey_ident")
dev.off()
```


```{r}
# Visualization
DimPlot(integrated, reduction = "umap", group.by = "pase_ident")
```































```{r}
# Create list containing new and old 48 h
# Split old
old_list <- SplitObject(old, split.by = "old.ident")
old_48 <- old_list[["10X_Embryoid"]]
# Split new
new_list <- SplitObject(new, split.by = "orig.ident")
new_48 <- new_list[["hESC_48h"]]
# Combine new and old
combined_48 <- list("old_48" = old_48, "new_48" = new_48)
```

```{r}
# Integration
# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = combined_48)
anchors <- FindIntegrationAnchors(object.list = combined_48,
anchor.features = features)
integrated_48 <- IntegrateData(anchorset = anchors)
```


```{r}
# Seperate new v. old
# Extract old.ident to a vector
batchname <- integrated_48$old.ident
# Rename
batchname <- gsub("10X_Embryoid", "old", batchname)
batchname <- gsub("hESC_48h", "new", batchname)
# Paste back to create a new metadata called batch
integrated_48$batch <- batchname
```

```{r}
# Integrated Analysis
DefaultAssay(integrated_48) <- "integrated"
integrated_48 <- ScaleData(integrated_48, verbose = FALSE)
integrated_48 <- RunPCA(integrated_48, npcs = 20, verbose = FALSE)
integrated_48 <- RunUMAP(integrated_48, reduction = "pca", dims = 1:20,
return.model = TRUE)
```

```{r}
# Integrated Clustering
integrated_48 <- FindNeighbors(integrated_48, reduction = "pca", dims = 1:20)
integrated_48 <- FindClusters(integrated_48, resolution = 0.2)
# Label Clusters
integrated_48 <- RenameIdents(integrated_48,
`0` = "Meso_2", `1` = "PGC",
`2` = "Meso_1", `3` = "AM")
```

```{r}
# Visualization
DimPlot(integrated_48, reduction = "umap")
```

```{r}
# Export
svg(file = "umap_cluster.svg", width = 6, height = 6)
DimPlot(integrated_48, reduction = "umap")
dev.off()
```

```{r}
# Dot plots
Idents(integrated_48) <- factor(Idents(integrated_48),
levels = c("Meso_2", "Meso_1", "AM", "PGC"))
markers_to_plot <- c("NANOG", "SOX2", "POU5F1", "TBXT", "MIXL1", "MESP1",
"EOMES", "GATA6", "APLNR", "LHX1", "PITX2", "PDGFRA",  "HEY1", "ISL1",
"TFAP2A", "GATA3", "ANXA3", "TCIM", "GABRP", "SOX17", "TFAP2C")
DotPlot(e12, cols = c("grey", "darkblue"),
features = markers_to_plot, dot.scale = 5) + RotatedAxis()
```

```{r}
# Export
svg(file = "dotplot.svg", width = 8, height = 4)
DotPlot(integrated_48, cols = c("grey", "darkblue"),
features = markers_to_plot, dot.scale = 5) + RotatedAxis()
dev.off()
```

```{r}
# Save data file processed thus far to .RDS
saveRDS(integrated_48, file = "new_old_48h_integrated.rds")
```

```{r}
# Read saved .RDS
integrated_48 <- readRDS("new_old_48h_integrated.rds")
```

```{r}
# Extract PCA as matrix
pca_data <- as.matrix(integrated_48$pca@cell.embeddings)

# Extract cell identity as a matrix
ident_data <- as.matrix(integrated_48@active.ident)
colnames(ident_data) <- "Cell Identity"

# Combine and export
pca_ident <- cbind(pca_data, ident_data)
write.csv(pca_ident, "pca_ident.csv")
```



```{r}
# Pre-processing - Human E14
# Extract gene name and id into separate columns in count matrix
gene_name <- strsplit(e14_mat$gene_id, "[|]")
gene_name_split <- as.data.frame(do.call(cbind, gene_name))
gene_id <- as.vector(t(gene_name_split[1, ]))
gene_name <- as.vector(t(gene_name_split[2, ]))
e14_mat$gene_id <- gene_id
e14_mat$gene_name <- gene_name

# Extract only gene id, reference and start/end location
e14_fpkm <- e14_fpkm[, c(2, 4, 6:7)]
colnames(e14_fpkm)[1] <- "gene_id"

# Merge fpkm with count matrix via id
e14_mat <- merge(e14_mat, e14_fpkm, by = "gene_id")
# Exclude any genes with reference other than "chr"
e14_mat <- e14_mat[grep("chr", e14_mat$Reference, fixed = FALSE), ]

# Rename columns with gene names (make names unique by adding ".1")
rownames(e14_mat) <- make.names(e14_mat$gene_name, unique = TRUE)
e14_mat$gene_id <- NULL
e14_mat$gene_name <- NULL
e14_mat[["Reference"]] <- NULL

# Normalize e14 data to 10X compatible format
e14_len <- e14_mat$End - e14_mat$Start
e14_mat[["Start"]] <- NULL
e14_mat[["End"]] <- NULL
e14_mat_norm <- e14_mat * 1000 / e14_len

# Retain d.p.f. information
days <- colnames(e14_mat_norm)
days[grep("D14", colnames(e14_mat_norm), fixed = FALSE)] <- "Day 14"
days[grep("D12", colnames(e14_mat_norm), fixed = FALSE)] <- "Day 12"
days[grep("D10", colnames(e14_mat_norm), fixed = FALSE)] <- "Day 10"
days[grep("D9", colnames(e14_mat_norm), fixed = FALSE)] <- "Day 9"
days[grep("D8", colnames(e14_mat_norm), fixed = FALSE)] <- "Day 8"
days[grep("D7", colnames(e14_mat_norm), fixed = FALSE)] <- "Day 7"
days[grep("D6", colnames(e14_mat_norm), fixed = FALSE)] <- "Day 6"

# Standard Seurat workflow
e14 <- CreateSeuratObject(counts = e14_mat_norm)
e14 <- NormalizeData(object = e14)
e14 <- FindVariableFeatures(object = e14)
# Cell Cycle Score
e14 <- CellCycleScoring(e14, s.features = s_genes,
g2m.features = g2m_genes, set.ident = TRUE)
e14[["CC.Difference"]] <- e14$S.Score - e14$G2M.Score
# Cluster data
e14 <- ScaleData(object = e14)
e14 <- RunPCA(object = e14)
e14 <- FindNeighbors(object = e14)
e14 <- FindClusters(object = e14, resolution = 0.2)
e14 <- RunUMAP(object = e14, reduction = "pca", dims = 1:20)
# Incorporate days information
e14$dpf <- days
# Subset epi and hypo
e14_epi_hypo <- subset(x = e14, idents = c("1", "5"))
# Rename clusters
cluster_id <- c("Epiblast", "Hypoblast")
names(cluster_id) <- levels(e14_epi_hypo)
e14_epi_hypo <- RenameIdents(e14_epi_hypo, cluster_id)
# Save meta data for integration analysis
e14_epi_hypo$e14_ident <- e14_epi_hypo@active.ident
# Create e14_epi
Idents(object = e14) <- "e14_ident"
e14_epi <- subset(e14_epi_hypo, idents = "Epiblast")
```


```{r}
```
